File: kakao.app/js/ked/edit/tool/cell.js
var _k_ = {empty: function (l) {return l==='' || l===null || l===undefined || l!==l || typeof(l) === 'object' && Object.keys(l).length === 0}, list: function (l) {return l != null ? typeof l.length === 'number' ? l : [] : []}, clamp: function (l,h,v) { var ll = Math.min(l,h), hh = Math.max(l,h); if (!_k_.isNum(v)) { v = ll }; if (v < ll) { v = ll }; if (v > hh) { v = hh }; if (!_k_.isNum(v)) { v = ll }; return v }, isNum: function (o) {return !isNaN(o) && !isNaN(parseFloat(o)) && (isFinite(o) || o === Infinity || o === -Infinity)}}

var cell


cell = (function ()
{
    function cell ()
    {}

    cell["cells"] = function (cols, rows)
    {
        return this.matrix(cols,rows,(function ()
        {
            return {bg:[],fg:[],char:' '}
        }))
    }

    cell["metas"] = function (cols, rows)
    {
        return this.matrix(cols,rows,(function ()
        {
            return {pre:[],pst:[]}
        }))
    }

    cell["matrix"] = function (cols, rows, cb)
    {
        var c, cells, l, lines

        lines = []
        for (var _a_ = l = 0, _b_ = rows; (_a_ <= _b_ ? l < rows : l > rows); (_a_ <= _b_ ? ++l : --l))
        {
            cells = []
            for (var _c_ = c = 0, _d_ = cols; (_c_ <= _d_ ? c < cols : c > cols); (_c_ <= _d_ ? ++c : --c))
            {
                cells.push(cb())
            }
            lines.push(cells)
        }
        return lines
    }

    cell["cellsForLines"] = function (lines)
    {
        var cells, width

        width = this.widthOfLines(lines)
        cells = this.cells(width,lines.length)
        this.stampLines(cells,lines)
        return cells
    }

    cell["cellSize"] = function (cells)
    {
        return [cells[0].length,cells.length]
    }

    cell["stampLines"] = function (cells, lines, x = 0, y = 0)
    {
        var char, ci, li, line

        if (_k_.empty(lines))
        {
            return
        }
        var _a_ = this.pos(x,y); x = _a_[0]; y = _a_[1]

        var list = _k_.list(lines)
        for (li = 0; li < list.length; li++)
        {
            line = list[li]
            var list1 = _k_.list(line)
            for (ci = 0; ci < list1.length; ci++)
            {
                char = list1[ci]
                cells[li][ci].char = char
            }
        }
    }

    cell["wrapCellRect"] = function (cells, x1, y1, x2, y2)
    {
        var cols, rows

        var _a_ = this.cellSize(cells); cols = _a_[0]; rows = _a_[1]

        if (x1 < 0)
        {
            x1 = cols + x1
        }
        if (x2 < 0)
        {
            x2 = cols + x2
        }
        if (y1 < 0)
        {
            y1 = rows + y1
        }
        if (y2 < 0)
        {
            y2 = rows + y2
        }
        return [x1,y1,x2,y2]
    }

    cell["clampCellRect"] = function (cells, x1, y1, x2, y2)
    {
        var cols, rows

        var _a_ = this.cellSize(cells); cols = _a_[0]; rows = _a_[1]

        x1 = _k_.clamp(0,cols - 1,x1)
        x2 = _k_.clamp(0,cols - 1,x2)
        y1 = _k_.clamp(0,rows - 1,y1)
        y2 = _k_.clamp(0,rows - 1,y2)
        return [x1,y1,x2,y2]
    }

    cell["cellsWithChar"] = function (cells, char)
    {
        var cell, res, row, x, y

        res = []
        var list = _k_.list(cells)
        for (y = 0; y < list.length; y++)
        {
            row = list[y]
            var list1 = _k_.list(row)
            for (x = 0; x < list1.length; x++)
            {
                cell = list1[x]
                if (cell.char === char)
                {
                    res.push({pos:[x,y],cell:cells[y][x]})
                }
            }
        }
        return res
    }

    cell["cellsInRect"] = function (cells, x1, y1, x2, y2)
    {
        var res, x, y

        var _a_ = this.clampCellRect(cells,x1,y1,x2,y2); x1 = _a_[0]; y1 = _a_[1]; x2 = _a_[2]; y2 = _a_[3]

        res = []
        for (var _b_ = y = y1, _c_ = y2; (_b_ <= _c_ ? y <= y2 : y >= y2); (_b_ <= _c_ ? ++y : --y))
        {
            for (var _d_ = x = x1, _e_ = x2; (_d_ <= _e_ ? x <= x2 : x >= x2); (_d_ <= _e_ ? ++x : --x))
            {
                res.push({pos:[x,y],cell:cells[y][x]})
            }
        }
        return res
    }

    cell["cellNeighborsAtPos"] = function (cells, x, y, xd = 1, yd = 1)
    {
        var x1, x2, y1, y2

        var _a_ = this.clampCellRect(cells,x - xd,y - yd,x + xd,y + yd); x1 = _a_[0]; y1 = _a_[1]; x2 = _a_[2]; y2 = _a_[3]

        return this.cellsInRect(cells,x1,y1,x2,y2)
    }

    return cell
})()

export default cell;

File: kakao.app/love/lua/edit/tool/cell.lua
--[[
     0000000  00000000  000      000      
    000       000       000      000      
    000       0000000   000      000      
    000       000       000      000      
     0000000  00000000  0000000  0000000  
--]]


local cell = class("cell")
    

--  0000000  00000000  000      000       0000000  
-- 000       000       000      000      000       
-- 000       0000000   000      000      0000000   
-- 000       000       000      000           000  
--  0000000  00000000  0000000  0000000  0000000   


function cell.static.cells(cols, rows) 
    return cell.matrix(cols, rows, (function () return {bg = array(), fg = array(), char = ' '} end))
    end


function cell.static.matrix(cols, rows, cb) 
        local lines = array()
        for l in iter(1, rows) do 
            local cells = array()
            for c in iter(1, cols) do 
                cells:push(cb())
            end
            
            lines:push(cells)
        end
        
        return lines
    end

-- 00000000   0000000   00000000       000      000  000   000  00000000   0000000  
-- 000       000   000  000   000      000      000  0000  000  000       000       
-- 000000    000   000  0000000        000      000  000 0 000  0000000   0000000   
-- 000       000   000  000   000      000      000  000  0000  000            000  
-- 000        0000000   000   000      0000000  000  000   000  00000000  0000000   


function cell.static.cellsForLines(lines) 
        local width = belt.widthOfLines(lines)
        local cells = cell.cells(width, #lines)
        
        cell.stampLines(cells, lines)
        
        return cells
    end

--  0000000  000  0000000  00000000  
-- 000       000     000   000       
-- 0000000   000    000    0000000   
--      000  000   000     000       
-- 0000000   000  0000000  00000000  


function cell.static.cellSize(cells) 
        if empty(cells) then 
            return 0, 0
        end
        
        return #cells[1], #cells
    end

--  0000000  000000000   0000000   00     00  00000000   
-- 000          000     000   000  000   000  000   000  
-- 0000000      000     000000000  000000000  00000000   
--      000     000     000   000  000 0 000  000        
-- 0000000      000     000   000  000   000  000        


function cell.static.stampLines(cells, lines, x, y) 
        x = x or 0
        y = y or 0
        
        if empty(lines) then return end
        
        local x, y = belt.pos(x, y)
        
        for li, line in ipairs(lines) do 
            for ci, char in ipairs(line) do 
                cells[li][ci].char = char
            end
        end
    end

-- 000   000  00000000    0000000   00000000   
-- 000 0 000  000   000  000   000  000   000  
-- 000000000  0000000    000000000  00000000   
-- 000   000  000   000  000   000  000        
-- 00     00  000   000  000   000  000        


function cell.static.wrapCellRect(cells, x1, y1, x2, y2) 
        local cols, rows = cell.cellSize(cells)
        
        if (x1 < 0) then x1 = (cols + x1) end
        if (x2 < 0) then x2 = (cols + x2) end
        
        if (y1 < 0) then y1 = (rows + y1) end
        if (y2 < 0) then y2 = (rows + y2) end
        
        return array(x1, y1, x2, y2)
    end

--  0000000  000       0000000   00     00  00000000   
-- 000       000      000   000  000   000  000   000  
-- 000       000      000000000  000000000  00000000   
-- 000       000      000   000  000 0 000  000        
--  0000000  0000000  000   000  000   000  000        


function cell.static.clampCellRect(cells, x1, y1, x2, y2) 
        local cols, rows = cell.cellSize(cells)
        
        x1 = clamp(1, cols, x1)
        x2 = clamp(1, cols, x2)
        
        y1 = clamp(1, rows, y1)
        y2 = clamp(1, rows, y2)
        
        return array(x1, y1, x2, y2)
    end

-- 000   000  000  000000000  000   000        0000000  000   000   0000000   00000000   
-- 000 0 000  000     000     000   000       000       000   000  000   000  000   000  
-- 000000000  000     000     000000000       000       000000000  000000000  0000000    
-- 000   000  000     000     000   000       000       000   000  000   000  000   000  
-- 00     00  000     000     000   000        0000000  000   000  000   000  000   000  


function cell.static.cellsWithChar(cells, char) 
        local res = array()
        for y, row in ipairs(cells) do 
            for x, cell in ipairs(row) do 
                if (cell.char == char) then 
                    res:push({pos = array(x, y), cell = cells[y][x]})
                end
            end
        end
        
        return res
    end

-- 000  000   000      00000000   00000000   0000000  000000000  
-- 000  0000  000      000   000  000       000          000     
-- 000  000 0 000      0000000    0000000   000          000     
-- 000  000  0000      000   000  000       000          000     
-- 000  000   000      000   000  00000000   0000000     000     


function cell.static.cellsInRect(cells, x1, y1, x2, y2) 
        local x1, y1, x2, y2 = unpack(cell.clampCellRect(cells, x1, y1, x2, y2))
        
        local res = {}
        for y in iter(y1, y2) do 
            for x in iter(x1, x2) do 
                res[(#res + 1)] = {pos = {x, y}, cell = cells[y][x]}
            end
        end
        
        return res
    end

-- 000   000  00000000  000   0000000   000   000  0000000     0000000   00000000    0000000  
-- 0000  000  000       000  000        000   000  000   000  000   000  000   000  000       
-- 000 0 000  0000000   000  000  0000  000000000  0000000    000   000  0000000    0000000   
-- 000  0000  000       000  000   000  000   000  000   000  000   000  000   000       000  
-- 000   000  00000000  000   0000000   000   000  0000000     0000000   000   000  0000000   


function cell.static.cellNeighborsAtPos(cells, x, y, xd, yd) 
        xd = xd or 1
        yd = yd or 1
        
        local x1, y1, x2, y2 = unpack(cell.clampCellRect(cells, (x - xd), (y - yd), (x + xd), (y + yd)))
        
        return cell.cellsInRect(cells, x1, y1, x2, y2)
    end

return cell